### PR TITLE
addClusterFacet

### PR DESCRIPTION
**Pull request by asthabh23**
## Description
This PR add `addCusterFacet` function and also add it's test coverage

## Related Issue
#21 

## How Has This Been Tested?
Tested on `cruncher.fixture.json` file which consists of real data

## Screenshots (if appropriate):
Screenshot 2024-12-03 at 18 18 14

For 92 URLs in the fixture, 1 cluster is created.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: distiller.js [hunk 1]
```diff
@@ -383,6 +384,61 @@ export class DataChunks {
...
+   * @param {object} clusterOptions options
+   * @param {number} clusterOptions.count number of clusters, The default value is log10(nValues)
+   * @param {function} clusterOptions.producer function that takes the cluster value and returns all possible cluster values
+   */
+  addClusterFacet(facetName, baseFacet, { count: clustercount = Math.floor(Math.log10(this.facets[baseFacet].length)) }) {
+    const facetValues = this.facets[baseFacet];
+
+    const createClusterMap = () => {
+      const clusterMap = facetValues.reduce((map, facet) => {
+        const clusters = urlProducer(facet.value);
```
Comment (trieloff) [inline distiller.js:403]: `urlProducer` should be `producer` and an argument to the function.

File: distiller.js [hunk 2]
```diff
@@ -383,6 +384,61 @@ export class DataChunks {
...
+   * will group the data into clusters based on the URL paths.
+   * You can specify the number of clusters and a producer function to
+   * generate the clusters.
+   * @param {string} facetName name of your new facet
+   * @param {string} baseFacet name of the base facet, from which to derive the clusters
+   * @param {object} clusterOptions options
+   * @param {number} clusterOptions.count number of clusters, The default value is log10(nValues)
+   * @param {function} clusterOptions.producer function that takes the cluster value and returns all possible cluster values
+   */
+  addClusterFacet(facetName, baseFacet, { count: clustercount = Math.floor(Math.log10(this.facets[baseFacet].length)) }) {
```
Comment (trieloff) [inline distiller.js:398]:
```suggestion
  addClusterFacet(facetName, baseFacet, { count: clustercount = Math.floor(Math.log10(this.facets[baseFacet].length)),
  producer: urlProducer }) {
```

File: distiller.js [hunk 3]
```diff
@@ -383,6 +384,61 @@ export class DataChunks {
...
+
+      for (const [cluster, count] of clusterMap.entries()) {
+        if (count > maxCount) {
+            maxCount = count;
+            mostOccurringCluster = cluster;
+        }
+      }
+
+      // Calculate the total number of URLs in the superset cluster
+      const totalUrlsInSupersetCluster = Math.floor(facetValues.length + clustercount);
```
Comment (trieloff) [inline distiller.js:425]:
```suggestion
      // Calculate the total number of items in the superset cluster
      const totalItemsInSupersetCluster = Math.floor(facetValues.length + clustercount);
```

Let's keep this generic here.

File: distiller.js [hunk 4]
```diff
@@ -383,6 +384,61 @@ export class DataChunks {
...
+    const { clusterMap, mostOccurringCluster, totalUrlsInSupersetCluster } = createClusterMap();
+    const sortedClusters = [...clusterMap.entries()]
+        .sort((a, b) => b[1] - a[1])
+        .slice(0, clustercount)
+        .map(([cluster]) => cluster);
+
+    this.addFacet(facetName, (bundle) => {
+        const facetMatch = facetValues.find(f => f.entries.some(e => e.id === bundle.id));
+        const clusters = urlProducer(facetMatch.value);
+        return { facetMatch, ...clusters.filter(cluster => sortedClusters.includes(cluster)) };
```
Comment (trieloff) [inline distiller.js:438]:
```suggestion
        return [ facetMatch, ...clusters.filter(cluster => sortedClusters.includes(cluster)) ];
```

facet functions return arrays, not objects.

**Review by trieloff — CHANGES_REQUESTED** [review]
Left some comments for you.

File: distiller.js [hunk 5]
```diff
@@ -383,6 +384,61 @@ export class DataChunks {
...
+      // Find the most occurring cluster
+      let mostOccurringCluster = null;
+      let maxCount = 0;
+
+      for (const [cluster, count] of clusterMap.entries()) {
+        if (count > maxCount) {
+            maxCount = count;
+            mostOccurringCluster = cluster;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (asthabh23) [reply distiller.js]: updated in [62ee81f](https://github.com/adobe/rum-distiller/pull/30/commits/62ee81f1bb8749a2907d6b605159bd0cadb22a59)

File: distiller.js [hunk 6]
```diff
@@ -383,6 +384,58 @@ export class DataChunks {
...
+        });
+        return map;
+      }, new Map());
+
+      // Find the most occurring cluster
+      let mostOccurringCluster = null;
+      const sortedClusters = [...clusterMap.entries()].sort((a, b) => b[1] - a[1]);
+      if (sortedClusters.length > 0) {
+        mostOccurringCluster = sortedClusters[0];
+      }
```
Comment (trieloff) [inline distiller.js:419]:
```suggestion
      const [mostOccurringCluster] = [...clusterMap.entries()].sort((a, b) => b[1] - a[1]);
```

this should be more concise.

**Review by trieloff — APPROVED** [review]
Almost there!

**Pull request merged**